core/src/main/java/org/elasticsearch/search/query/QueryPhase.java

```diff
@@ -361,6 +361,15 @@ public TopDocs call() throws Exception {
                 }
             }
 
+            if (collector != null) {
+                final Collector child = collector;
+                collector = new CancellableCollector(searchContext, collector);
```

Comment:
This part can cause significant slow downs. There may be ways that we can make it better, eg. by doing it at the bulk scorer level so that we can perform the check less often. Can we remove it for now and work on making this part cancellable in another PR. Or at least make it an opt-in for now with a setting?